src/filesystem_littlefs.c

```diff
@@ -0,0 +1,517 @@
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    oldName = process_path(oldName, devoptab_littlefs.name);
+    newName = process_path(newName, devoptab_littlefs.name);
+
+    result = lfs_rename(lfs, oldName, newName);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int chmod_r(struct _reent *r, const char *path, mode_t mode) {
+    r->_errno = ENOTSUP;
+    return -1;
+}
+
+static int fchmod_r(struct _reent *r, void *fileStruct, mode_t mode) {
+    r->_errno = ENOTSUP;
+    return -1;
+}
+
+static int mkdir_r(struct _reent *r, const char *path, int mode) {
+    int result;
+
+    if (IS_EMPTY(path)) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    path = process_path(path, devoptab_littlefs.name);
+
+    result = lfs_mkdir(lfs, path);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static DIR_ITER *diropen_r(struct _reent *r, DIR_ITER *dirState, const char *path) {
+    int result;
+
+    if (IS_EMPTY(path)) {
+        r->_errno = EINVAL;
+        return NULL;
+    }
+
+    path = process_path(path, devoptab_littlefs.name);
+
+    result = lfs_dir_open(lfs, dirState->dirStruct, path);
+    r->_errno = get_errno(result);
+
+    if (result < 0) {
+        return NULL;
+    }
+
+    return dirState;
+}
+
+static int dirreset_r(struct _reent *r, DIR_ITER *dirState) {
+    int result;
+
+    result = lfs_dir_rewind(lfs, dirState->dirStruct);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int dirnext_r(struct _reent *r, DIR_ITER *dirState, char *filename, struct stat *filestat) {
+    int result;
+    struct lfs_info info;
+
+    if (filename == NULL || filestat == NULL) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    result = lfs_dir_read(lfs, dirState->dirStruct, &info);
+    r->_errno = get_errno(result);
+
+    if (result == 0) {
+        // End of the directory
+        r->_errno = ENOENT; // has to be set at the end
+        return -1;
+    } else if (result < 0) {
+        return -1;
+    }
+
+    if (info.type & LFS_TYPE_DIR) {
+        filestat->st_mode = S_IFDIR;
+    } else if (info.type & LFS_TYPE_REG) {
+        filestat->st_mode = S_IFREG;
+    } else {
+        // Unexpected type, skip to the next dir
+        return dirnext_r(r, dirState, filename, filestat);
+    }
+
+    strncpy(filename, info.name, NAME_MAX);
+
+    return 0;
+}
+
+static int dirclose_r(struct _reent *r, DIR_ITER *dirState) {
+    int result;
+
+    result = lfs_dir_close(lfs, dirState->dirStruct);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int statvfs_r(struct _reent *r, const char *path, struct statvfs *buf) {
+    if (IS_EMPTY(path)) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    memset(buf, 0, sizeof(struct statvfs));
+
+    buf->f_frsize = lfs->cfg->block_size;
+    buf->f_bfree = lfs->free.size;
+    buf->f_bavail = buf->f_bfree;
+    buf->f_files = 0; // TODO: Count all inodes
```

Comment:
Do we need the number of inodes yet for something? Or is it just a nice-to-have option?

Reply:
Not really. It could be used maybe to check if we can create another file for this filesystem but I don't expect we will get even close to the limit.